onnx/optimizer/passes/fuse_bn_into_conv.h

```diff
@@ -0,0 +1,212 @@
+    auto s_iter = graph.getInitializer(bn_inputs[1]->uniqueName());
+    auto bbn_iter = graph.getInitializer(bn_inputs[2]->uniqueName());
+    auto m_iter = graph.getInitializer(bn_inputs[3]->uniqueName());
+    auto var_iter = graph.getInitializer(bn_inputs[4]->uniqueName());
+    auto W_iter = graph.getInitializer(conv_inputs[1]->uniqueName());
+    if (s_iter == end_iter || bbn_iter == end_iter || m_iter == end_iter ||
+        var_iter == end_iter || W_iter == end_iter) {
+      return false;
+    }
+
+    ONNX_ASSERT(s_iter->sizes().size() == 1);
+    ONNX_ASSERT(bbn_iter->sizes().size() == 1 && bbn_iter->sizes()[0] == s_iter->sizes()[0]);
+    ONNX_ASSERT(m_iter->sizes().size() == 1 && m_iter->sizes()[0] == s_iter->sizes()[0]);
+    ONNX_ASSERT(var_iter->sizes().size() == 1 && var_iter->sizes()[0] == s_iter->sizes()[0]);
+    ONNX_ASSERT(W_iter->sizes().size() > 2 && W_iter->sizes()[0] == s_iter->sizes()[0]);
+    ONNX_ASSERT(s_iter->elem_type() == bbn_iter->elem_type() &&
+                s_iter->elem_type() == m_iter->elem_type() &&
+                s_iter->elem_type() == var_iter->elem_type() &&
+                s_iter->elem_type() == W_iter->elem_type());
+    if (s_iter->elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT &&
+        s_iter->elem_type() != ONNX_NAMESPACE::TensorProto_DataType_DOUBLE)  {
+      return false;
+    }
+
+    auto s = *s_iter;
+    auto bbn = *bbn_iter;
+    auto m = *m_iter;
+    auto var = *var_iter;
+    auto W = *W_iter;
+    auto epsilon = bn->hasAttribute(kepsilon) ? bn->f(kepsilon) : 1e-5;
+    Tensor eps;
+    Tensor bc;
+
+    switch(s.elem_type()) {
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT: {
+        eps.sizes().push_back(s.sizes()[0]);
+        eps.elem_type() = ONNX_NAMESPACE::TensorProto_DataType_FLOAT;
+        for (int64_t i = 0; i < eps.sizes()[0]; ++i)  {
+          eps.floats().push_back((float) epsilon);
+        }
+        bc.sizes().push_back(s.sizes()[0]);
+        bc.elem_type() = ONNX_NAMESPACE::TensorProto_DataType_FLOAT;
+        for (int64_t i = 0; i < eps.sizes()[0]; ++i)  {
+          bc.floats().push_back(0.f);
+        }
+
+        if (conv_inputs.size() == 3) {
+          auto bc_iter = graph.getInitializer(conv_inputs[2]->uniqueName());
+          if (bc_iter == end_iter) {
+            return false;
+          }
+          bc = *bc_iter;
+          ONNX_ASSERT(bc.sizes().size() == 1 && bc.sizes()[0] == s.sizes()[0]);
+        }
+        var.add(eps);
+        var.sqrt();
+        s.divide(var);
+
+        if (!s.is_raw_data()) {
+          const char * ptr = reinterpret_cast<const char *>(s.floats().data());
+          std::string string_rep(ptr, ptr + sizeof(float) * s.floats().size());
+          s.set_raw_data(string_rep);
+        }
+        W.scale_by_first_dim(s);
+        bc.subtract(m);
+        bc.multiply(s);
+        bc.add(bbn);
+        break;
+      }
+      case ONNX_NAMESPACE::TensorProto_DataType_DOUBLE:
```

Comment:
these two cases are almost identical, use another macro to avoid duplication?

Reply:
used macro